secret-manager/runtime/src/main/java/io/quarkiverse/googlecloudservices/secretmanager/runtime/config/SecretManagerConfigSourceFactory.java

```diff
@@ -0,0 +1,26 @@
+package io.quarkiverse.googlecloudservices.secretmanager.runtime.config;
+
+import java.util.Collections;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import com.google.cloud.ServiceOptions;
+
+import io.smallrye.config.ConfigSourceContext;
+import io.smallrye.config.ConfigSourceFactory;
+
+public class SecretManagerConfigSourceFactory implements ConfigSourceFactory {
+
+    private static final String PROJECT_ID_PROPERTY = "quarkus.google.cloud.project-id";
+
+    @Override
+    public Iterable<ConfigSource> getConfigSources(ConfigSourceContext configSourceContext) {
+        String projectId = configSourceContext.getValue(PROJECT_ID_PROPERTY).getValue();
+        if (projectId == null) {
+            // Fallback to the default project detected via Application Default Credentials.
+            projectId = ServiceOptions.getDefaultProjectId();
```

Comment:
This is interesting, I wonder if we can use the same trick on the various other extensions that uses the project ID.
How this is working ? Would it load the current project from a compute instance or a function ?

Reply:
Yes, this detects the default GCP project configured for authentication in the environment.

For example, if you run the code on Cloud Run, the Cloud Run instance sets some environment variables so you can authenticate in the libraries automatically via "Application Default Credentials" ([more info here](https://cloud.google.com/docs/authentication/production#automatically)). (It will use the default service account configured if no additional auth settings are provided).

Or if you develop locally and set the `GOOGLE_APPLICATION_CREDENTIALS` manually or via the [gcloud SDK](https://cloud.google.com/sdk/gcloud/reference/auth/application-default), the `ServiceOptions.getDefaultProjectId()` derives this setting from the environment.

Yes indeed -- you could have the `quarkus.google.cloud.project-id` set in the common module to use `ServiceOptions.getDefaultProjectId()` by default. And then users may decide to override this value by setting it manually if they want.

Reply:
Great, this could provide the same functionalities than #4 without needed to deal with the metadata server manually right ?

Reply:
Ahh, yes. I think so. I'll send a PR out with what I have in mind; it should be able to solve that problem.